### ISSUE TITLE
Autofac 4.0

### ISSUE THREAD

**Issue by janus007** [NancyFx/Nancy.Bootstrappers.Autofac#25]
Hello

I was trying to get AutofacNancyBootstrapper to work with Autofac 4.0 because I wanted to use .NET 4.5.2 in Visual Studio 2015, but with no luck.

It seems that AutofacNancyBootstrapper utilizes ILifetimeScope and that appears not to work with Autofac 4.0

So my question... when will I be able to use Autofac 4.0?

Thank you for a great product.

**Comment by khellang** [comment]
Uuuh... `ILifetimeScope` is kinda *the* core interface in Autofac. It would be weird if they dropped it in 4.0. I'll check and come back to you.

**Comment by janus007** [comment]
Hi khellang
That was quick reply :) They didn't drop it, I can see that my previous post could indicate that. 

The case is that I'm trying to use ASP.NET 4.5.2 and AutofacRegistration.Populate or builder.Populate, either way this extension method exists in Autofac 4.* and as soon I reference that, the AutofacNancyBootstrapper complains about the Autofac version. 

Normally I would just do an assembly-binding, but that is not possible in vNext, it is being said that assembly-version is carried out by Roslyn, but apparently this is not working. Don't now why, but I guess it has something to do with the target framework that Nancy is built in, at the moment 4.0.

**Comment by khellang** [comment]
Right, so this has nothing to do with .NET 4.5.2, but rather that you're trying to use this in a DNX project? We don't support DNX yet. We have a branch going in the main repo to get it running (in fact I think it runs on DNX451, but not on DNXCORE50). Before we ship that, we'll make sure to have a version of the different bootstrappers that works with it.

**Comment by janus007** [comment]
Amazing... I will look forward to compatibility with 451.

**Comment by radams0x** [comment]
Hello, 
I'm having the same issue trying to use dnx451. I'll keep an eye out for your updated bootstrapper. Thanks!

**Comment by khellang** [comment]
If you really want to use Nancy with Autofac on DNX451, you can simply copy the bootstrapper over as source, temporarily. I've created a sample repo over at https://github.com/khellang/Nancy.Autofac.Dnx. That runs nicely with Nancy on DNX451 with Autofac in harmony with IServiceCollection and co :grin:

**Comment by radams0x** [comment]
Wow, thank you so much for this! I was actually just in the process of doing that, but but this helped me as I couldn't get it to compile and hadn't noticed the need for the ComponentContextExtensions.cs.
I'll give it a whirl and post the results.
Thanks!!

**Comment by khellang** [comment]
No problem :+1: BTW, you don't **need** Autofac 4.0 to run on DNX. You need it if you want to use `Autofac.Framework.DependencyInjection`, which contains the `Populate` method to populate the `ContainerBuilder` using an `IServiceCollection`. If you don't need services from DNX, you can get by with simply using the `Nancy.Bootstrappers.Autofac` package.

**Comment by radams0x** [comment]
Yes, I'm using Autofac.Framework.DependencyInjection, as described to use the Populate method to populate the ContainerBuilder using an IServiceCollection.

I tried the code from the sample you just posted.
Here's what I get:
"Nancy: Could not initialize bootstrapper of type '<my namespace>.Startup+Bootstrapper'.\n   at Nancy.Bootstrapper.NancyBootstrapperLocator.LocateBootstrapper()\r\n   at Nancy.Bootstrapper.NancyBootstrapperLocator.get_Bootstrapper()\r\n   at Nancy.Owin.NancyOptions.get_Bootstrapper()\r\n   at Nancy.Owin.NancyMiddleware.UseNancy(NancyOptions options)\r\n   at Nancy.Owin.DelegateExtensions.UseNancy(Action`1 builder, NancyOptions options)\r\n   at Nancy.Owin.DelegateExtensions.UseNancy(Action`1 builder, Action`1 action)\r\n   at <my namespace>.Startup.<>c.<Configure>b__12_0(Action`1 x) in <path>\\Startup.cs:line 141\r\n   at Microsoft.AspNet.Builder.OwinExtensions.UseOwin(IApplicationBuilder builder, Action`1 pipeline)\r\n   at <my namespace>.Startup.Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerFactory loggerFactory) in <path>\\Startup.cs:line 141"

**Comment by radams0x** [comment]
HI khellang,

I got it to work as you describe, use the code you posted. at https://github.com/khellang/Nancy.Autofac.Dnx

I am using Autofac.Framework.DependencyInjection, which contains the Populate method to populate the ContainerBuilder using an IServiceCollection.

Thank you so much for everything!!

**Comment by janus007** [comment]
Haven't tested it, but I believe radams0x is right on.

khellang, I thank you very much for this :)

**Comment by khellang** [comment]
Glad to hear :+1: I'll leave this issue open until we get a permanent solution :smile:

**Comment by norgie** [comment]
Any chance of having the readme.md file updated? There's currently no mentioning of AutofacNancyBootstrapper (or the other IoC bootstrappers) being abstract forcing us to implement GetAllModules, GetModule and RegisterModules.

**Comment by khellang** [comment]
@norgie What are you talking about? You shouldn't have to implement anything. All those methods you mention are implemented by the `AutofacNancyBootstrapper`.

**Comment by norgie** [comment]
Right now I'm trying out Ninject, but I get the same error with Autofac and Structuremap, i.e.: Abstract inherited member 'IEnumerable<INancyModule> Nancy.Bootstrapper.NancyBootstrapperBase<IKernel>.GetAllModules(NancyContext) is not implemented' The same goes for the two other methods.

**Comment by norgie** [comment]
@khellang Seems like it is Resharper that's confused. I get a red squigly and the message I mentioned above when hovering over the squigly. No problem compiling the project. Sorry for raising a false alarm.

**Comment by khellang** [comment]
Heh. What version of R# are you using? Some EAP?

**Comment by norgie** [comment]
VS2012, R# 7.1.3 Full version. Latest Nancy, latest TopShelf.

**Comment by norgie** [comment]
@khellang 
![r cockup](https://cloud.githubusercontent.com/assets/1003903/13916101/ec9fcdb2-ef58-11e5-910f-df656631d8e2.png)

**Comment by norgie** [comment]
@khellang Closing the solution and opening it again seems to get rid of the squigly. Weird.
